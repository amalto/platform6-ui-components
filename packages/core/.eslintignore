dist/
loader/

src/components.d.ts