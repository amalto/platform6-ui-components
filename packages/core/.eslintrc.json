{
  "extends": ["../../.eslintrc.json", "plugin:@stencil/recommended", "plugin:jest/recommended", "plugin:jsx-a11y/recommended"],
  "parserOptions": {
    "project": ["./tsconfig.json", "./tsconfig.lint.json"],
    "ecmaFeatures": {
      "jsx": true // Allows for the parsing of JSX
    }
  },
  "plugins": ["jest"],

  "rules": {
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "assert": "either"
      }
    ],
    "react/jsx-props-no-spreading": "off",
    "react/no-unknown-property": [2, { "ignore": ["class"] }]
  },
  "settings": {
    "react": {
      "pragma": "h"
    }
  }
}
