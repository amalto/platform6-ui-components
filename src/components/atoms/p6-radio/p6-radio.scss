:host {
  display: inline-block;
  font-family: $family-primary;

  input,
  label {
    cursor: pointer;
  }

  input {
    position: absolute;
    opacity: 0;
  }

  label {
    cursor: pointer;
    display: inline-block;
    font-size: $size-normal;
    line-height: 13px;
    margin-right: $gap-size * 1px;
    position: relative;
    user-select: none;
    vertical-align: top;
    width: auto;

    &::before {
      content: '';
      background-color: transparent;
      border: 1px solid $input-border-color;
      border-radius: 50px;
      box-shadow: inset 0 0 0 2px $input-background-color;
      display: inline-block;
      height: 10px;
      margin-right: 5px;
      padding: 5px;
      vertical-align: top;
      width: 10px;
    }
  }

  input:checked + label::before {
    background-color: $success;
  }

  input[readonly]:checked + label::before,
  input[disabled]:checked + label::before {
    background-color: $input-disabled-color;
    cursor: not-allowed;
  }

  input[readonly] + label,
  input[disabled] + label {
    color: $input-disabled-color;
    cursor: not-allowed;
  }
}
