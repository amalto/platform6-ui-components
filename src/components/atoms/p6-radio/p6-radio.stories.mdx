import {
  Story,
  Preview,
  Meta,
  Markdown,
  Description,
} from "@storybook/addon-docs/blocks";

import Readme from "./readme.md";

import {
  DefaultStory,
  ReadonlyStory,
  DisabledStory,
  FormStory,
} from "./p6-radio.stories.ts";

<Meta title="Atoms/Radio" component="p6-radio" />

# Radio

<Preview>
  <Story
    name="Default"
    height="84px"
    parameters={DefaultStory.parameters}
    args={DefaultStory.args}
    argTypes={DefaultStory.argTypes}
  >
    {DefaultStory}
  </Story>
</Preview>

<Preview>
  <Story
    name="Readonly"
    height="84px"
    parameters={ReadonlyStory.parameters}
    args={ReadonlyStory.args}
    argTypes={ReadonlyStory.argTypes}
  >
    {ReadonlyStory}
  </Story>
</Preview>

<Preview>
  <Story
    name="Disabled"
    height="84px"
    parameters={DisabledStory.parameters}
    args={DisabledStory.args}
    argTypes={DisabledStory.argTypes}
  >
    {DisabledStory}
  </Story>
</Preview>

<Preview>
  <Story
    name="Form"
    height="84px"
    parameters={FormStory.parameters}
    args={FormStory.args}
    argTypes={FormStory.argTypes}
  >
    {FormStory}
  </Story>
</Preview>

<Description markdown={Readme} />
