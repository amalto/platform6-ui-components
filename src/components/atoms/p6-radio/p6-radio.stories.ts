/* eslint-disable import/no-extraneous-dependencies */
import { makeStory } from "../../../shared/storybook/makeStory";
/* eslint-enable import/no-extraneous-dependencies */
import { getComponent, getForm, Prop } from "../../../shared/storybook/stories";

const getInput = (text: string, props?: Prop): string => {
  return getComponent("p6-radio", text, props);
};

export const DefaultStory = makeStory<{
  disabled: boolean;
  readOnly: boolean;
  checked: boolean;
  name: string;
  label: string;
  value: string;
}>({
  args: {
    disabled: false,
    readOnly: false,
    checked: false,
    name: "field",
    label: "Label",
    value: "value",
  },
  builder: ({ label, ...args }): string => getInput(label, { ...args }),
});

export const ReadonlyStory = makeStory({
  builder: (): string =>
    getInput("Read only", {
      name: "readonly-example",
      checked: true,
      readOnly: true,
    }),
});

export const DisabledStory = makeStory({
  builder: (): string =>
    getInput("Disabled", {
      name: "disabled-example",
      checked: true,
      disabled: true,
    }),
});

export const FormStory = makeStory({
  builder: (): string =>
    getForm(
      ["first", "second", "third"]
        .map((key) => getInput(key, { value: key, name: "field" }))
        .join("\n")
    ),
});
