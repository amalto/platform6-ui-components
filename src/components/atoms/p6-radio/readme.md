# p6-radio

<!-- Auto Generated Below -->

## Properties

| Property             | Attribute  | Description   | Type               | Default     |
| -------------------- | ---------- | ------------- | ------------------ | ----------- |
| `checked`            | `checked`  | Initial value | `boolean`          | `false`     |
| `disabled`           | `disabled` | Disable       | `boolean`          | `false`     |
| `name` _(required)_  | `name`     | Radio name    | `string`           | `undefined` |
| `readOnly`           | `readonly` | Readonly      | `boolean`          | `false`     |
| `value` _(required)_ | `value`    | Value         | `number \| string` | `undefined` |

## Events

| Event              | Description                                                                            | Type                                                           |
| ------------------ | -------------------------------------------------------------------------------------- | -------------------------------------------------------------- |
| `p6FormRegister`   | Registering the field in a p6-form                                                     | `CustomEvent<P6Control<string \| number \| undefined>>`        |
| `p6FormUnregister` | Unregistering the field in a p6-form                                                   | `CustomEvent<P6Control<string \| number \| undefined>>`        |
| `p6Invalid`        | Fires when the field has been checked for validity and doesn't satisfy its constraints | `CustomEvent<InvalidEventDetail>`                              |
| `p6Valid`          | Fires when the field has been checked for validity and satisfy its constraints         | `CustomEvent<ValidEventDetail<string \| number \| undefined>>` |

## Methods

### `checkValidity() => Promise<boolean>`

Returns whether a form will validate when it is submitted, without having to submit it.

#### Returns

Type: `Promise<boolean>`

### `reset() => Promise<boolean>`

Restores the radio's default value

#### Returns

Type: `Promise<boolean>`

### `validationMessage() => Promise<string>`

Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
It also triggers the standard error message, such as "this is a required field".

#### Returns

Type: `Promise<string>`

---

_Built with [StencilJS](https://stenciljs.com/)_
