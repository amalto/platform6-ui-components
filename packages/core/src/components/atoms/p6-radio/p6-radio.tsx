import { Component, ComponentInterface, Element, Event, EventEmitter, h, Host, JSX, Method, Prop } from '@stencil/core';
import { P6Control } from '../../../shared/form/control';
import { InvalidEventDetail, ValidEventDetail } from '../../../shared/form/event';
import { defaultCheckValidity, defaultValidationMessage } from '../../../shared/form/validation';

export type P6RadioValue = string | number | undefined;

@Component({
  tag: 'p6-radio',
  styleUrl: './p6-radio.scss',
  shadow: true,
})
export class P6Radio implements ComponentInterface, P6Control<P6RadioValue> {
  @Element() host!: HTMLP6RadioElement;

  /**
   * Radio name
   */
  @Prop({ reflect: true }) name!: string;

  /**
   * Value
   */
  @Prop() value!: string | number;

  /**
   * Initial value
   */
  // eslint-disable-next-line @stencil/strict-mutable
  @Prop({ mutable: true, reflect: true }) checked = false;

  /**
   * Disable
   */
  @Prop() disabled = false;

  /**
   * Readonly
   */
  @Prop({ attribute: 'readOnly' }) readOnly = false;

  /**
   * Registering the field in a p6-form
   */
  @Event() p6FormRegister!: EventEmitter<P6Control<P6RadioValue>>;

  /**
   * Unregistering the field in a p6-form
   */
  @Event() p6FormUnregister!: EventEmitter<P6Control<P6RadioValue>>;

  /**
   * Fires when the field has been checked for validity and satisfy its constraints
   */
  @Event() p6Valid!: EventEmitter<ValidEventDetail<P6RadioValue>>;

  /**
   * Fires when the field has been checked for validity and doesn't satisfy its constraints
   */
  @Event() p6Invalid!: EventEmitter<InvalidEventDetail>;

  private nativeInput: HTMLInputElement | undefined;

  private defaultValue: boolean | undefined;

  componentWillLoad(): void {
    this.host.addEventListener('focusout', this.checkValidity.bind(this));
  }

  render(): JSX.Element {
    const { host, name, checked, disabled, readOnly, value } = this;
    const inputId = `${name}-${value}-input`;

    return (
      <Host class={host?.className}>
        <input checked={checked} id={inputId} disabled={disabled} name={name} readOnly={readOnly} type="radio" value={value} onClick={this.updateState} />
        <label htmlFor={inputId}>
          <slot />
        </label>
      </Host>
    );
  }

  componentDidLoad(): void {
    if (this.defaultValue === undefined) {
      this.defaultValue = this.nativeInput?.checked ?? this.checked;
    }

    this.p6FormRegister.emit(this);
  }

  disconnectedCallback?(): void {
    this.p6FormUnregister.emit(this);
  }

  /**
   * Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
   * It also triggers the standard error message, such as "this is a required field".
   */
  @Method()
  async validationMessage(): Promise<string> {
    return defaultValidationMessage(this.nativeInput);
  }

  /**
   * Returns whether a form will validate when it is submitted, without having to submit it.
   */
  @Method()
  async checkValidity(): Promise<boolean> {
    return defaultCheckValidity<P6RadioValue>({
      name: this.name,
      disabled: this.disabled,
      nativeInput: this.nativeInput,
      p6Valid: this.p6Valid,
      p6Invalid: this.p6Invalid,
      validationMessage: this.validationMessage.bind(this),
      getValue: () => {
        return this.nativeInput?.checked ? this.nativeInput?.value : undefined;
      },
    });
  }

  /**
   * Restores the radio's default value
   */
  @Method()
  async reset(): Promise<boolean> {
    this.checked = !!this.defaultValue;
    return Promise.resolve(true);
  }

  private updateState = (): void => {
    this.checked = !this.checked;
  };
}
