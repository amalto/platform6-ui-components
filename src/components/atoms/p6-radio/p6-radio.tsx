import { Component, Element, h, Host, Prop, State } from "@stencil/core";

@Component({
  tag: "p6-radio",
  styleUrl: "./p6-radio.scss",
  scoped: true,
})
export class P6Radio {
  @Element() host?: HTMLP6RadioElement;

  /**
   * Radio name
   */
  @Prop() name!: string;

  /**
   * Initial value
   */
  @Prop() checked = false;

  /**
   * Disable
   */
  @Prop() disabled = false;

  /**
   * Readonly
   */
  @Prop() readonly = false;

  /**
   * Value
   */
  @Prop() value!: string | number;

  /**
   * State of the radio
   */
  @State() isChecked = false;

  private clickHandler = (event: Event): void => {
    if (this.readonly) {
      event.preventDefault();
      return;
    }
    this.isChecked = !this.isChecked;
  };

  componentWillLoad(): void {
    this.isChecked = this.checked;
  }

  render(): JSX.Element {
    const { host, name, isChecked, disabled, readonly, value } = this;

    const inputId = `${name}-${value}-input`;

    return (
      <Host class={host?.className}>
        <input
          checked={isChecked}
          id={inputId}
          disabled={disabled}
          name={name}
          onClick={this.clickHandler}
          readOnly={readonly}
          type="radio"
          value={value}
        />
        <label htmlFor={inputId}>
          <slot />
        </label>
      </Host>
    );
  }
}
