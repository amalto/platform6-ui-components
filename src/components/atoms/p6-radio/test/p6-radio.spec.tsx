import { newSpecPage } from '@stencil/core/testing';
import { P6Radio } from '../p6-radio';

describe('p6-radio', () => {

  const name: string = 'radio-input';
  const value: string = name;
  const extraClass: string = 'mgr-5';
  const id: string = `${name}-${value}-input`;

  it('defaut p6-radio', async () => {
    const page = await newSpecPage({
      components: [P6Radio],
      html: `
      <p6-radio name="${name}"
        class="${extraClass}" value="${value}">
        first
      </p6-radio>
      `,
    });
    expect(page.root).toEqualHtml(`
      <p6-radio name="${name}" class="${extraClass}" value="${value}">
        <input id="${id}"
          name="${name}"
          type="radio"
          value="${value}"
        />
        <label htmlFor="${id}">
          first
        </label>
      </p6-radio>
    `);
  });
});
