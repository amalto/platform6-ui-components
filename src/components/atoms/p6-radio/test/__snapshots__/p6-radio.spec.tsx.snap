// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-radio defaut p6-radio 1`] = `
<p6-radio class="mgr-5" name="radio-input" value="radio-input">
  <mock:shadow-root>
    <input id="radio-input-radio-input-input" name="radio-input" type="radio" value="radio-input">
    <label htmlfor="radio-input-radio-input-input">
      <slot></slot>
    </label>
  </mock:shadow-root>
  first
</p6-radio>
`;
